### ISSUE TITLE
MSELoss (and others) ignore Reduction

### ISSUE THREAD

**Issue by IAmMahiro** [cool-japan/torsh#47]
The issue is simple...

In ```crates/torsh-nn/src/functional.rs```
Losses just ignore Reduction:

```rust
/// Mean Squared Error Loss
pub struct MSELoss {
    pub reduction: Reduction,
}

impl MSELoss {
    pub fn new(reduction: Reduction) -> Self {
        Self { reduction }
    }
}

impl CustomLoss for MSELoss {
    fn forward(&self, predictions: &Tensor, targets: &Tensor) -> Result<Tensor> {
        crate::functional::loss::mse_loss(predictions, targets, "none")
    }

    fn reduction(&self) -> &Reduction {
        &self.reduction
    }
}
```

```rust
/// Negative Log Likelihood Loss
pub struct NLLLoss {
    pub weight: Option<Tensor>,
    pub ignore_index: Option<i64>,
    pub reduction: Reduction,
}

impl NLLLoss {
    pub fn new(weight: Option<Tensor>, ignore_index: Option<i64>, reduction: Reduction) -> Self {
        Self {
            weight,
            ignore_index,
            reduction,
        }
    }
}

impl CustomLoss for NLLLoss {
    fn forward(&self, predictions: &Tensor, targets: &Tensor) -> Result<Tensor> {
        crate::functional::loss::nll_loss(
            predictions,
            &targets.cast_i64()?,
            self.weight.as_ref(),
            self.ignore_index,
            "none",
        )
    }

    fn reduction(&self) -> &Reduction {
        &self.reduction
    }
}
```

and others

**Comment by cool-japan** [comment]
Thanks @IAmMahiro. The hardcoded `"none"` is deliberate — but you have found a real bug next to it, and I would like to check whether that is the symptom you actually hit.

**Why the `"none"` is correct.** `CustomLoss` provides a wrapper that applies the reduction after `forward` (`crates/torsh-nn/src/functional/loss_advanced.rs:92-98`):

```rust
fn compute_loss(&self, predictions: &Tensor, targets: &Tensor) -> Result<Tensor> {
    self.validate_inputs(predictions, targets)?;
    let raw_loss = self.forward(predictions, targets)?;
    let batch_size = predictions.shape().dims()[0];
    self.reduction().apply(&raw_loss, batch_size)
}
```

So `forward` is contractually the *unreduced, element-wise* kernel and `compute_loss` is the public entry point. All 20 `CustomLoss` impls follow that contract, every documented call site uses `compute_loss` (`crates/torsh-nn/docs/CUSTOM_MODULE_TUTORIAL.md:877`), and all 25 assertions in `crates/torsh-nn/tests/hardening_nn_losses.rs` do too, pinning None, Mean and Sum to three distinct values. The composites depend on this: `WeightedLoss` (`:218`), `CombinedLoss` (`:273`) and `AdaptiveLoss` (`:318`) call the inner `forward` precisely to get element-wise values before applying their own reduction.

That said, the contract is nowhere documented on the trait, which is a fair thing to be caught by — I will add that doc line.

**The real defect.** `Reduction::Mean` divides by the batch size, not the element count (`loss_advanced.rs:63-65`, fed `predictions.shape().dims()[0]` from line 93), whereas the functional API in the same crate divides by `numel` (`crates/torsh-nn/src/functional/loss.rs:1051` → `loss.mean(None, false)`). So these two disagree:

```rust
MSELoss::new(Reduction::Mean).compute_loss(&x, &y)?   // sum / batch
F::mse_loss(&x, &y, "mean")?                          // sum / numel
```

by a factor equal to the number of non-batch elements per sample. The second matches PyTorch. The divisor logic is in fact byte-identical to the `"batchmean"` arm at `loss.rs:497-502` — so `Reduction::Mean` is currently `batchmean`.

It is visible in the pinned tests: `hardening_nn_losses.rs:152` uses `DIMS = [2, 3]` (numel 6, batch 2), and SmoothL1 is pinned at `Sum = 2.46125`, `Mean = 1.230625` — that is Sum/2, not Sum/6.

That divergence is confirmed by arithmetic on the checked-in fixtures, independently of how the issue was found — so it stands on its own. If you did hit it in a real training run rather than while reading the source, say so, since that bears on priority.

Decided fix: make `Reduction::Mean` divide by `numel`, matching PyTorch and the functional layer, and add an explicit `Reduction::BatchMean` for the current behaviour. This changes loss magnitudes and is accepted as a breaking change — it will carry a CHANGELOG entry, and the `*/Mean` values pinned in `hardening_nn_losses.rs` will be recomputed by hand rather than re-baselined.

Two smaller things found alongside, tracked separately: `IoULoss::forward` (`loss_advanced.rs:175-196`) returns an already-global `[1]` tensor and so breaks the element-wise `forward` contract the other 19 impls honour, and the trait default `fn reduction(&self) -> &Reduction { &Reduction::Mean }` (`:69-71`) means third-party impls silently inherit the batchmean semantics.

All tracked in `TODO.md` under "GitHub Issue Triage Findings" and targeted for the **0.2.1** release. Leaving this issue open, retargeted onto the divisor bug — thanks for pointing at the right file even though the mechanism turned out to be one layer over.
